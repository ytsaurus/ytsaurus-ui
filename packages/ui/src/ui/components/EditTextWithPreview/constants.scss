$tabbed-content-header-height: 48px;
