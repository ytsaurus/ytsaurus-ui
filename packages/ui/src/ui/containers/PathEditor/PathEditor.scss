@import '../../styles/mixins';

.path-editor {
    position: relative;

    display: block;

    &__items {
        @include reset-ul();

        max-height: 300px;

        overflow: hidden auto;

        border: 1px solid var(--dark-divider);
        border-top: none;
        box-shadow: 0 10px 20px -5px var(--opacity-background);
        background-color: var(--main-background);
    }

    &__item {
        @include reset-li();
        @include ellipsis();

        line-height: 30px;

        width: 100%;
        padding: 0 20px;

        vertical-align: bottom;

        color: var(--primary-text);
        cursor: pointer;

        &:hover {
            background-color: var(--hover-background);
        }

        &_error {
            color: var(--danger-color);
        }

        &-path {
            margin-left: 12px;
        }

        &_selected_yes {
            background-color: var(--hover-background);
        }
    }

    &__popup {
        z-index: var(--dialog-suggest-z-index) !important;
    }
}
