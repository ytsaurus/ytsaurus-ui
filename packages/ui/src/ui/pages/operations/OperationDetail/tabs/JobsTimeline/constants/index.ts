export const ROW_HEIGHT = 30;
export const TIMELINE_HEIGHT = 12;
export const TIMELINE_SELECT_BORDER_WIDTH = 2;
export const OFFSET = (ROW_HEIGHT - TIMELINE_HEIGHT) / 2;
export const MAX_JOBS_COUNT = 10000;
