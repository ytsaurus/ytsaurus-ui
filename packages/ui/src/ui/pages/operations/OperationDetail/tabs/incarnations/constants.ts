import b from 'bem-cn-lite';

const block = b('incarnations');

export const detailsCn = block('details');
export const failReasonsCn = block('fail-reasons');
export const incarnationCn = block('incarnation');
export const incarnationInfoCn = block('incarnation-info');
export const incarnationButtonCn = block('incarnation-button');
export const incarnationsCountTypeCn = block('incarnations-count', {type: true});
export const incarnationsCountItemCn = block('incarnations-count', {item: true});
export const incarnationsCountItemCountCn = block('incarnations-count', {count: true});
