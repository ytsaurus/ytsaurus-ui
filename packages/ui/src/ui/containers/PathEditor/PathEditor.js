import React, {Component} from 'react';
import {connect} from 'react-redux';
import ReactList from 'react-list';
import PropTypes from 'prop-types';
import block from 'bem-cn-lite';
import key from 'hotkeys-js';

import debounce_ from 'lodash/debounce';
import find_ from 'lodash/find';
import isEmpty_ from 'lodash/isEmpty';

import {Popup, TextInput} from '@gravity-ui/uikit';

import Icon from '../../components/Icon/Icon';
import ErrorMessage from '../../components/ErrorMessage/ErrorMessage';

import {
    filterByCurrentPath,
    getCompletedPath,
    getIconNameForType,
    getNextSelectedIndex,
    getPrevSelectedIndex,
} from '../../utils/navigation/path-editor';
import {
    loadSuggestionsList,
    removeActiveRequests,
} from '../../store/actions/navigation/path-editor/path-editor';
import {KeyCode} from '../../constants/index';

import './PathEditor.scss';

const debounceTime = 300;
const b = block('path-editor');

export class PathEditor extends Component {
    static propTypes = {
        // from connect
        suggestions: PropTypes.arrayOf(
            PropTypes.shape({
                path: PropTypes.string.isRequired,
                targetPathBroken: PropTypes.bool,
                type: PropTypes.string,
            }),
        ).isRequired,
        suggestionsError: PropTypes.bool,
        errorMessage: PropTypes.string,

        loadSuggestionsList: PropTypes.func.isRequired,
        removeActiveRequests: PropTypes.func.isRequired,
        // from parent component
        className: PropTypes.string,
        placeholder: PropTypes.string,
        customFilter: PropTypes.func,
        onChange: PropTypes.func,
        onFocus: PropTypes.func,
        onBlur: PropTypes.func,
        onApply: PropTypes.func,
        onCancel: PropTypes.func,
        defaultPath: PropTypes.string,
        disabled: PropTypes.bool,
        autoFocus: PropTypes.bool,
        hasClear: PropTypes.bool,
        showErrors: PropTypes.bool,
    };

    static defaultProps = {
        errorMessage: 'Oops, something went wrong',
        placeholder: 'Enter the path...',
        suggestionsError: false,
        autoFocus: false,
        showErrors: true,
        defaultPath: undefined,
        disabled: false,
        hasClear: false,
    };

    constructor(props) {
        super(props);

        this.suggestionsList = React.createRef();
        this.input = React.createRef();

        this.debounceLoading = debounce_(this.debounceLoading, debounceTime);
        this.prevScope = null;

        this.state = {
            path: props.defaultPath,
            actualSuggestions: [],
            inputFocus: false,
            inputChange: false,
            selectedIndex: -1,
        };
    }

    static getDerivedStateFromProps(props, state) {
        const res = {};
        if (state.inputFocus && state.inputChange) {
            Object.assign(res, {
                actualSuggestions: props.suggestions.length
                    ? filterByCurrentPath(state.path, props.suggestions)
                    : [],
            });
        }

        if (state.path === undefined && props.defaultPath !== undefined) {
            Object.assign(res, {
                path: props.defaultPath,
            });
        }

        return isEmpty_(res) ? null : res;
    }

    componentDidMount() {
        const {loadSuggestionsList, customFilter, autoFocus} = this.props;
        const {path} = this.state;

        this.prevScope = key.getScope();
        if (path) {
            loadSuggestionsList(path, customFilter);
        }
        if (autoFocus) {
            this._setFocus();
        }
    }

    componentDidUpdate(prevProps) {
        if (prevProps.disabled && !this.props.disabled && this.input.current) {
            this._setFocus();
        }
    }

    componentWillUnmount() {
        this.props.removeActiveRequests();
    }

    _setFocus() {
        this.input.current.focus();
    }

    get inputWidth() {
        // Is there a way to avoid using private _control property?
        return this.input.current && this.input.current.offsetWidth - 2;
    }

    callCallback(cb, ...params) {
        if (typeof cb === 'function') {
            return cb(...params);
        }
    }

    debounceLoading(path) {
        const {loadSuggestionsList, customFilter, onChange} = this.props;

        loadSuggestionsList(path, customFilter);
        this.callCallback(onChange, path);
    }

    hideSuggestions = () => {
        this.setState({inputFocus: false, selectedIndex: -1});
    };

    selectPrevSuggestion() {
        const {selectedIndex, actualSuggestions} = this.state;

        const prevIndex = getPrevSelectedIndex(actualSuggestions, selectedIndex);
        this.setState({selectedIndex: prevIndex});
    }

    selectNextSuggestion() {
        const {selectedIndex, actualSuggestions} = this.state;

        const nextIndex = getNextSelectedIndex(actualSuggestions, selectedIndex);
        this.setState({selectedIndex: nextIndex});
    }

    handleInputChange = (path) => {
        this.setState({path, selectedIndex: -1, inputChange: true});
        this.debounceLoading(path);
    };

    handleInputFocus = () => {
        const {onFocus} = this.props;
        const {path} = this.state;

        key.setScope('path-editor');
        this.setState({inputFocus: true});
        this.callCallback(onFocus, path);
    };

    handleInputBlur = () => {
        const {onBlur} = this.props;
        const {path} = this.state;

        key.setScope(this.prevScope);
        this.hideSuggestions();
        this.callCallback(onBlur, path);
    };

    handleEnterClick(evt) {
        evt.preventDefault();

        const {selectedIndex, actualSuggestions} = this.state;
        const {onApply} = this.props;
        const inputPath = evt.target.value;

        if (selectedIndex === -1) {
            this.setState({path: inputPath, selectedIndex: -1});
            this.callCallback(onApply, inputPath);
        } else {
            const suggestion = find_(actualSuggestions, (item, index) => index === selectedIndex);
            const completedPath = getCompletedPath(suggestion);

            this.handleInputChange(completedPath);
        }
    }

    handleEscClick() {
        const {onCancel} = this.props;

        this.input.current.blur();
        this.callCallback(onCancel);
    }

    handleTabClick(evt) {
        evt.preventDefault();
        const {actualSuggestions} = this.state;

        if (actualSuggestions.length === 1) {
            const completedPath = getCompletedPath(actualSuggestions[0]);

            this.handleInputChange(completedPath);
        } else if (actualSuggestions.length > 1) {
            this.selectNextSuggestion();
        }
    }

    handleKeyDown = (evt) => {
        const key = evt.keyCode;

        switch (key) {
            case KeyCode.ARROW_DOWN:
                this.selectNextSuggestion();
                break;
            case KeyCode.ARROW_UP:
                this.selectPrevSuggestion();
                break;
            case KeyCode.ENTER:
                this.handleEnterClick(evt);
                break;
            case KeyCode.ESCAPE:
                this.handleEscClick();
                break;
            case KeyCode.TAB:
                this.handleTabClick(evt);
                break;
        }
    };

    renderInput() {
        const {placeholder, autoFocus, hasClear, disabled} = this.props;
        const {path} = this.state;

        return (
            <TextInput
                onKeyDown={this.handleKeyDown}
                onUpdate={this.handleInputChange}
                onFocus={this.handleInputFocus}
                onBlur={this.handleInputBlur}
                placeholder={placeholder}
                hasClear={hasClear}
                autoFocus={autoFocus}
                disabled={disabled}
                controlRef={this.input}
                value={path}
            />
        );
    }

    renderItem = (index, key) => {
        const {selectedIndex, actualSuggestions} = this.state;

        const item = actualSuggestions[index];
        const {type, dynamic} = item;
        const iconType = type === 'table' && dynamic ? 'table_dynamic' : type;
        const completedPath = getCompletedPath(item);
        const isSelected = index === selectedIndex ? 'yes' : 'no';
        const iconName = getIconNameForType(iconType, item.targetPathBroken);

        const mouseDownHandler = (evt) => {
            this.handleInputChange(completedPath);
            evt.preventDefault();
        };

        return (
            <div
                key={key}
                onMouseDown={mouseDownHandler}
                className={b('item', {selected: isSelected})}
            >
                <Icon awesome={iconName} />

                <span className={b('item-path')}>{item.path}</span>
            </div>
        );
    };

    renderSuggestions() {
        const {actualSuggestions} = this.state;

        return actualSuggestions.length ? (
            <ReactList
                itemRenderer={this.renderItem}
                length={actualSuggestions.length}
                type="simple"
            />
        ) : null;
    }

    renderError() {
        const {errorMessage} = this.props;

        return <ErrorMessage className={b('item', {error: true})} message={errorMessage} />;
    }

    renderPopup() {
        const {suggestionsError, showErrors} = this.props;
        const {actualSuggestions, inputFocus} = this.state;

        const width = this.inputWidth || 0;
        const isVisible =
            (actualSuggestions.length || (suggestionsError && showErrors)) && inputFocus;
        const content =
            suggestionsError && showErrors ? this.renderError() : this.renderSuggestions();

        return (
            <Popup
                className={b('popup')}
                placement={['bottom-start', 'top-start']}
                onClose={this.hideSuggestions}
                anchorRef={this.input}
                open={isVisible}
                offset={[undefined, 0]}
            >
                <div className={b('items')} style={{width}} ref={this.suggestionsList}>
                    {content}
                </div>
            </Popup>
        );
    }

    render() {
        return (
            <div className={b(null, this.props.className)}>
                {this.renderInput()}
                {this.renderPopup()}
            </div>
        );
    }
}

const mapStateToProps = ({navigation}) => ({
    suggestions: navigation.pathEditor.suggestions,
    suggestionsError: navigation.pathEditor.suggestionsError,
    errorMessage: navigation.pathEditor.errorMessage,
});

export default connect(mapStateToProps, {
    loadSuggestionsList,
    removeActiveRequests,
})(PathEditor);
