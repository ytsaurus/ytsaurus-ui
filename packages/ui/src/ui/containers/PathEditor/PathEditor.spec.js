import 'jsdom-global/register';

import React from 'react';
import Enzyme, {mount, shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import _ from 'lodash';

import {PathEditor} from './PathEditor';

import {
    filterByCurrentPath,
    getCompletedPath,
    getNextSelectedIndex,
    getPrevSelectedIndex,
} from '../../utils/navigation/path-editor';
import {KeyCode} from '../../constants/index';

Enzyme.configure({adapter: new Adapter()});

jest.mock('../../utils/navigation/path-editor', () => ({
    filterByCurrentPath: jest.fn(),
    getNextSelectedIndex: jest.fn(),
    getPrevSelectedIndex: jest.fn(),
    getCompletedPath: jest.fn(),
    getIconNameForType: jest.fn(),
}));

jest.mock('../../store/actions/navigation/path-editor/path-editor', () => ({
    loadSuggestionsList: jest.fn(),
    removeActiveRequests: jest.fn(),
}));
jest.mock('../../common/hammer');

const INPUT_WIDTH = 200;

const props = {
    suggestions: [],
    defaultPath: '//path/to/object',
    placeholder: 'placeholder',
    disabled: false,
    loadSuggestionsList: jest.fn(),
    removeActiveRequests: jest.fn(),
    onChange: jest.fn(),
    onFocus: jest.fn(),
    onBlur: jest.fn(),
    onApply: jest.fn(),
    onCancel: jest.fn(),
    autoFocus: true,
};
const suggestions = [
    {type: 'map_node', path: 'path/to/node1'},
    {type: 'map_node', path: 'path/to/node2'},
    {type: 'map_node', path: 'path/to/node3'},
];

describe('<PathEditor />', () => {
    let wrapper;

    beforeEach(() => {
        jest.spyOn(PathEditor.prototype, 'inputWidth', 'get').mockImplementation(() => INPUT_WIDTH);
        jest.spyOn(PathEditor.prototype, '_setFocus').mockImplementation();

        jest.spyOn(PathEditor.prototype, 'selectNextSuggestion');
        jest.spyOn(PathEditor.prototype, 'selectPrevSuggestion');

        jest.spyOn(PathEditor.prototype, 'renderSuggestions');
        jest.spyOn(PathEditor.prototype, 'handleEnterClick');
        jest.spyOn(PathEditor.prototype, 'handleEscClick');
        jest.spyOn(PathEditor.prototype, 'handleTabClick');
        jest.spyOn(PathEditor.prototype, 'callCallback');
        jest.spyOn(PathEditor.prototype, 'renderError');
        jest.spyOn(PathEditor.prototype, 'renderInput');

        jest.spyOn(PathEditor, 'getDerivedStateFromProps');

        jest.spyOn(_, 'debounce').mockImplementation((fn) => fn);

        wrapper = shallow(<PathEditor {...props} />);

        wrapper.instance().input.current = {
            focus: jest.fn(),
            blur: jest.fn(),
        };
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render PathEditor container', () => {
        expect(wrapper.find('.path-editor').exists()).toBeTrue();
        expect(wrapper.find('.path-editor').children().length).toBe(2);
    });

    it('should put the path into the state from the props', () => {
        const pathFromState = wrapper.state().path;
        const pathFromProps = wrapper.instance().props.defaultPath;

        expect(pathFromState).toBe(pathFromProps);
    });

    it('should not call renderError() method if there are no errors', () => {
        expect(wrapper.instance().props.showErrors).toBeTrue();
        expect(wrapper.instance().props.suggestionsError).toBeFalse();
        expect(PathEditor.prototype.renderError).not.toHaveBeenCalled();
    });

    it('should call renderError() method once errors are present', () => {
        wrapper.setProps({...props, suggestionsError: true});
        expect(PathEditor.prototype.renderError).toHaveBeenCalledTimes(1);
    });

    describe('getDerivedStateFromProps', () => {
        it('should return null from getDerivedStateFromProps() if input has not been changed or focused', () => {
            expect(PathEditor.getDerivedStateFromProps).toHaveReturnedWith(null);
        });

        it('should return object from getDerivedStateFromProps() method if input has been changed and focused', () => {
            filterByCurrentPath.mockImplementation((path, suggestions) =>
                suggestions.filter((item, index) => index % 2 !== 0),
            );

            wrapper.setState({inputFocus: true, inputChange: true});

            expect(PathEditor.getDerivedStateFromProps).toHaveBeenCalledTimes(2);
            expect(filterByCurrentPath).not.toHaveBeenCalled();
            expect(PathEditor.getDerivedStateFromProps).toHaveReturnedWith({
                actualSuggestions: [],
            });

            wrapper.setProps({suggestions});

            expect(PathEditor.getDerivedStateFromProps).toHaveBeenCalledTimes(3);
            expect(filterByCurrentPath).toHaveBeenCalledTimes(1);
            expect(filterByCurrentPath).toHaveBeenCalledWith(wrapper.state().path, suggestions);

            expect(PathEditor.getDerivedStateFromProps).toHaveReturnedWith({
                actualSuggestions: [
                    {
                        type: 'map_node',
                        path: 'path/to/node2',
                    },
                ],
            });
        });
    });

    describe('componentDidMount', () => {
        it('should call loadSuggestionsList() action-creator after render', () => {
            const path = wrapper.state().path;
            const customFilter = wrapper.instance().props.customFilter;

            expect(props.loadSuggestionsList).toHaveBeenCalledTimes(1);
            expect(props.loadSuggestionsList).toHaveBeenCalledWith(path, customFilter);
        });

        it('should call _setFocus() method after render', () => {
            expect(PathEditor.prototype._setFocus).toHaveBeenCalledTimes(1);
        });
    });

    describe('componentDidUpdate', () => {
        it('should call _setFocus() method after update', () => {
            wrapper.setProps({disabled: true});

            expect(PathEditor.prototype._setFocus).toHaveBeenCalledTimes(1); // after componentDidMount

            wrapper.setProps({disabled: false});

            expect(PathEditor.prototype._setFocus).toHaveBeenCalledTimes(2);
        });
    });

    describe('componentWillUnmount', () => {
        it('should abort the request', () => {
            wrapper.unmount();

            expect(props.removeActiveRequests).toHaveBeenCalled();
        });
    });

    describe('Input', () => {
        it('should call renderInput() method once', () => {
            expect(PathEditor.prototype.renderInput).toHaveBeenCalledTimes(1);
        });

        it('should update state after the input has been changed', () => {
            const newPath = props.defaultPath + '1';

            wrapper.instance().handleInputChange(newPath);

            const {selectedIndex, path, inputChange} = wrapper.state();

            expect(selectedIndex).toBe(-1);
            expect(path).toBe(newPath);
            expect(inputChange).toBeTrue();
        });

        it('should call debounceLoading() method after input has been changed', () => {
            jest.spyOn(wrapper.instance(), 'debounceLoading');
            const newPath = props.defaultPath + '1';
            const debounceLoading = wrapper.instance().debounceLoading;

            expect(debounceLoading).not.toHaveBeenCalled();

            wrapper.instance().handleInputChange(newPath);

            expect(debounceLoading).toHaveBeenCalledTimes(1);
            expect(debounceLoading).toHaveBeenCalledWith(newPath);
        });

        it('should call loadSuggestionsList() and callCallback() methods after input has been changed', () => {
            const newPath = props.defaultPath + '1';
            const loadSuggestionsList = props.loadSuggestionsList;
            const callCallback = PathEditor.prototype.callCallback;

            expect(loadSuggestionsList).toHaveBeenCalledTimes(1); // after componentDidMount
            expect(loadSuggestionsList).toHaveBeenCalledWith(props.defaultPath, props.customFilter);
            expect(callCallback).not.toHaveBeenCalled();

            wrapper.instance().handleInputChange(newPath);

            expect(loadSuggestionsList).toHaveBeenCalledTimes(2);
            expect(loadSuggestionsList).toHaveBeenCalledWith(newPath, props.customFilter);

            expect(callCallback).toHaveBeenCalledTimes(1);
            expect(callCallback).toHaveBeenCalledWith(props.onChange, newPath);
        });

        it('should update the state after the input has been focused', () => {
            expect(wrapper.state().inputFocus).toBeFalse();

            wrapper.instance().handleInputFocus();

            expect(wrapper.state().inputFocus).toBeTrue();
        });

        it('should call callCallback() method after the input has been focused', () => {
            const callCallback = PathEditor.prototype.callCallback;

            expect(callCallback).not.toHaveBeenCalled();

            wrapper.instance().handleInputFocus();

            expect(callCallback).toHaveBeenCalledWith(props.onFocus, wrapper.state().path);
        });

        it('should update the state after the input has been blurred', () => {
            jest.spyOn(wrapper.instance(), 'hideSuggestions');
            expect(wrapper.state().inputFocus).toBeFalse();

            wrapper.instance().handleInputFocus();
            wrapper.instance().handleInputBlur();

            expect(wrapper.state().inputFocus).toBeFalse();
        });

        it('should call callCallback() method after the input has been blurred', () => {
            const callCallback = PathEditor.prototype.callCallback;

            expect(callCallback).not.toHaveBeenCalled();

            wrapper.instance().handleInputBlur();

            expect(callCallback).toHaveBeenCalledWith(props.onBlur, wrapper.state().path);
        });

        it('should pass the correct props into the Input component', () => {
            let input = wrapper.find('.path-editor').children().get(0);
            const newPath = props.defaultPath + '1';
            const {path} = wrapper.state();

            expect(input.props.value).toBe(path);
            expect(input.props.placeholder).toBe(props.placeholder);
            expect(input.props.disabled).toBe(props.disabled);
            expect(input.props.onUpdate).toBe(wrapper.instance().handleInputChange);
            expect(input.props.onFocus).toBe(wrapper.instance().handleInputFocus);
            expect(input.props.onBlur).toBe(wrapper.instance().handleInputBlur);

            wrapper.instance().handleInputChange(newPath);
            input = wrapper.find('.path-editor').children().get(0);

            expect(input.props.value).toBe(newPath);

            wrapper.setProps({
                disabled: true,
                placeholder: 'new placeholder',
            });
            input = wrapper.find('.path-editor').children().get(0);

            expect(input.props.placeholder).toBe('new placeholder');
            expect(input.props.disabled).toBeTrue();
        });
    });

    describe('events handling', () => {
        const getEvent = (keyCode) => ({
            keyCode,
            target: {value: '//path/to/another/folder'},
            preventDefault() {},
        });

        it('should handle the arrowDown event', () => {
            const selectNextSuggestion = PathEditor.prototype.selectNextSuggestion;
            const arrowDownEvt = getEvent(KeyCode.ARROW_DOWN);

            expect(selectNextSuggestion).not.toHaveBeenCalled();

            wrapper.instance().handleKeyDown(arrowDownEvt);

            expect(selectNextSuggestion).toHaveBeenCalledTimes(1);
        });

        it('should handle the arrowUp event', () => {
            const selectPrevSuggestion = PathEditor.prototype.selectPrevSuggestion;
            const arrowUpEvt = getEvent(KeyCode.ARROW_UP);

            expect(selectPrevSuggestion).not.toHaveBeenCalled();

            wrapper.instance().handleKeyDown(arrowUpEvt);

            expect(selectPrevSuggestion).toHaveBeenCalledTimes(1);
        });

        it('should update state after selectPrevSuggestion() method has been called', () => {
            const prevSelectedIndex = 2;
            const arrowUpEvt = getEvent(KeyCode.ARROW_UP);
            const {actualSuggestions, selectedIndex} = wrapper.state();

            expect(getPrevSelectedIndex).not.toHaveBeenCalled();

            getPrevSelectedIndex.mockImplementation(() => prevSelectedIndex);
            wrapper.instance().handleKeyDown(arrowUpEvt);

            expect(getPrevSelectedIndex).toHaveBeenCalledTimes(1);
            expect(getPrevSelectedIndex).toHaveBeenCalledWith(actualSuggestions, selectedIndex);
            expect(wrapper.state().selectedIndex).toBe(prevSelectedIndex);
        });

        it('should handle the enter event', () => {
            const handleEnterClick = PathEditor.prototype.handleEnterClick;
            const enterEvt = getEvent(KeyCode.ENTER);

            expect(handleEnterClick).not.toHaveBeenCalled();

            wrapper.instance().handleKeyDown(enterEvt);

            expect(handleEnterClick).toHaveBeenCalledTimes(1);
            expect(handleEnterClick).toHaveBeenCalledWith(enterEvt);
        });

        it('should update the state and call callCallback() method after enter key has been clicked', () => {
            const enterEvt = getEvent(KeyCode.ENTER);
            const newPath = enterEvt.target.value;

            wrapper.instance().handleKeyDown(enterEvt);

            expect(wrapper.state().selectedIndex).toBe(-1);
            expect(wrapper.state().path).toBe(newPath);
            expect(PathEditor.prototype.callCallback).toBeCalledWith(props.onApply, newPath);
        });

        it('should call handleInputChange() method after enter key has been clicked', () => {
            jest.spyOn(wrapper.instance(), 'handleInputChange');

            const enterEvt = getEvent(KeyCode.ENTER);
            const newPath = '//completed/path/to/folder/';
            const selectedIndex = 5;

            wrapper.setState({selectedIndex});
            getCompletedPath.mockImplementation(() => newPath);
            wrapper.instance().handleKeyDown(enterEvt);

            expect(wrapper.instance().handleInputChange).toHaveBeenCalledTimes(1);
            expect(wrapper.instance().handleInputChange).toHaveBeenCalledWith(newPath);
        });

        it('should handle the esc event', () => {
            const handleEscClick = PathEditor.prototype.handleEscClick;
            const escEvt = getEvent(KeyCode.ESCAPE);

            expect(handleEscClick).not.toHaveBeenCalled();

            wrapper.instance().handleKeyDown(escEvt);

            expect(handleEscClick).toHaveBeenCalledTimes(1);
        });

        it('should call callCallback() method after Esc key has been clicked', () => {
            const escEvt = getEvent(KeyCode.ESCAPE);

            wrapper.instance().handleKeyDown(escEvt);

            expect(PathEditor.prototype.callCallback).toBeCalledWith(props.onCancel);
        });

        it('should handle the tab event', () => {
            const handleTabClick = PathEditor.prototype.handleTabClick;
            const tabEvt = getEvent(KeyCode.TAB);

            expect(handleTabClick).not.toHaveBeenCalled();

            wrapper.instance().handleKeyDown(tabEvt);

            expect(handleTabClick).toHaveBeenCalledTimes(1);
            expect(handleTabClick).toHaveBeenCalledWith(tabEvt);
        });

        it('should call handleInputChange() method after tab key has been clicked', () => {
            jest.spyOn(wrapper.instance(), 'handleInputChange');

            const tabEvt = getEvent(KeyCode.TAB);
            const path = '//path/to/folder';
            const suggestion = {type: 'map_node', path};
            const actualSuggestions = [suggestion];

            wrapper.setState({actualSuggestions});
            getCompletedPath.mockImplementation((suggestion) => suggestion.path);
            wrapper.instance().handleKeyDown(tabEvt);

            expect(wrapper.instance().handleInputChange).toHaveBeenCalledTimes(1);
            expect(wrapper.instance().handleInputChange).toHaveBeenCalledWith(path);
        });

        it('should call selectNextSuggestion() method after tab key has been clicked', () => {
            const selectNextSuggestion = PathEditor.prototype.selectNextSuggestion;
            const tabEvt = getEvent(KeyCode.TAB);
            const actualSuggestions = [
                {type: 'map_node', path: '//path/to/folder1'},
                {type: 'map_node', path: '//path/to/folder2'},
            ];

            wrapper.setState({actualSuggestions});
            wrapper.instance().handleKeyDown(tabEvt);

            expect(selectNextSuggestion).toHaveBeenCalledTimes(1);
        });

        it('should update state after selectNextSuggestion() method has been called', () => {
            const nextSelectedIndex = 1;
            const arrowDownEvt = getEvent(KeyCode.ARROW_DOWN);
            const {actualSuggestions, selectedIndex} = wrapper.state();

            expect(getNextSelectedIndex).not.toHaveBeenCalled();

            getNextSelectedIndex.mockImplementation(() => nextSelectedIndex);
            wrapper.instance().handleKeyDown(arrowDownEvt);

            expect(getNextSelectedIndex).toHaveBeenCalledTimes(1);
            expect(getNextSelectedIndex).toHaveBeenCalledWith(actualSuggestions, selectedIndex);
            expect(wrapper.state().selectedIndex).toBe(nextSelectedIndex);
        });
    });

    describe('snapshots', () => {
        it('should match snapshot from shallow rendering', () => {
            const wrapper = shallow(<PathEditor {...props} />);

            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it('should match snapshot from mount rendering without suggestions', () => {
            const wrapper = mount(<PathEditor {...props} />);

            expect(toJson(wrapper)).toMatchSnapshot();

            wrapper.unmount();
        });

        // FIXME: it is don't work.. why?
        // it('should match snapshot from mount rendering with suggestions', () => {
        //     const wrapper = mount(<PathEditor {...props} suggestions={suggestions}/>);
        //     const actualSuggestions = [{type: 'map_node', path: '//path/to/folder/'}];
        //
        //     wrapper.setState({ actualSuggestions })
        //     expect(toJson(wrapper)).toMatchSnapshot();
        //     console.log(wrapper.debug());
        //
        //     wrapper.unmount();
        // });
    });
});
