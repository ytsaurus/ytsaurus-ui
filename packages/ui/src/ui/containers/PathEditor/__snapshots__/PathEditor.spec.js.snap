// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PathEditor /> snapshots should match snapshot from mount rendering without suggestions 1`] = `
<PathEditor
  autoFocus={true}
  defaultPath="//path/to/object"
  disabled={false}
  errorMessage="Oops, something went wrong"
  hasClear={false}
  loadSuggestionsList={
    [MockFunction] {
      "calls": Array [
        Array [
          "//path/to/object",
          undefined,
        ],
        Array [
          "//path/to/object",
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onApply={[MockFunction]}
  onBlur={[MockFunction]}
  onCancel={[MockFunction]}
  onChange={[MockFunction]}
  onFocus={[MockFunction]}
  placeholder="placeholder"
  removeActiveRequests={[MockFunction]}
  showErrors={true}
  suggestions={Array []}
  suggestionsError={false}
>
  <div
    className="path-editor"
  >
    <ForwardRef(TextInput)
      autoFocus={true}
      controlRef={
        Object {
          "current": <input
            class="yc-text-input__control yc-text-input__control_type_input"
            placeholder="placeholder"
            value="//path/to/object"
          />,
        }
      }
      disabled={false}
      hasClear={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onUpdate={[Function]}
      placeholder="placeholder"
      value="//path/to/object"
    >
      <span
        className="yc-text-input yc-text-input_view_normal yc-text-input_size_m yc-text-input_pin_round-round"
      >
        <InputControl
          autoFocus={true}
          controlProps={
            Object {
              "style": Object {},
            }
          }
          controlRef={[Function]}
          disabled={false}
          hasClear={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onUpdate={[Function]}
          placeholder="placeholder"
          value="//path/to/object"
        >
          <input
            autoFocus={true}
            className="yc-text-input__control yc-text-input__control_type_input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="placeholder"
            style={Object {}}
            value="//path/to/object"
          />
        </InputControl>
      </span>
    </ForwardRef(TextInput)>
    <Popup
      anchorRef={
        Object {
          "current": <input
            class="yc-text-input__control yc-text-input__control_type_input"
            placeholder="placeholder"
            value="//path/to/object"
          />,
        }
      }
      className="path-editor__popup"
      offset={
        Array [
          undefined,
          0,
        ]
      }
      onClose={[Function]}
      open={false}
      placement={
        Array [
          "bottom-start",
          "top-start",
        ]
      }
    >
      <Portal>
        <Portal
          containerInfo={
            <body>
              <div
                class="yc-toaster"
              />
            </body>
          }
        >
          <CSSTransition
            addEndListener={[Function]}
            appear={true}
            classNames={
              Object {
                "appear": "yc-popup-wrapper_appear",
                "appearActive": "yc-popup-wrapper_appear_active",
                "appearDone": "yc-popup-wrapper_appear_done",
                "enter": "yc-popup-wrapper_enter",
                "enterActive": "yc-popup-wrapper_enter_active",
                "enterDone": "yc-popup-wrapper_enter_done",
                "exit": "yc-popup-wrapper_exit",
                "exitActive": "yc-popup-wrapper_exit_active",
                "exitDone": "yc-popup-wrapper_exit_done",
              }
            }
            in={false}
            mountOnEnter={true}
            nodeRef={
              Object {
                "current": null,
              }
            }
            unmountOnExit={true}
          >
            <Transition
              addEndListener={[Function]}
              appear={true}
              enter={true}
              exit={true}
              in={false}
              mountOnEnter={true}
              nodeRef={
                Object {
                  "current": null,
                }
              }
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              unmountOnExit={true}
            />
          </CSSTransition>
        </Portal>
      </Portal>
    </Popup>
  </div>
</PathEditor>
`;

exports[`<PathEditor /> snapshots should match snapshot from shallow rendering 1`] = `
<div
  className="path-editor"
>
  <ForwardRef(TextInput)
    autoFocus={true}
    controlRef={
      Object {
        "current": null,
      }
    }
    disabled={false}
    hasClear={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onUpdate={[Function]}
    placeholder="placeholder"
    value="//path/to/object"
  />
  <Popup
    anchorRef={
      Object {
        "current": null,
      }
    }
    className="path-editor__popup"
    offset={
      Array [
        undefined,
        0,
      ]
    }
    onClose={[Function]}
    open={false}
    placement={
      Array [
        "bottom-start",
        "top-start",
      ]
    }
  >
    <div
      className="path-editor__items"
      style={
        Object {
          "width": 200,
        }
      }
    />
  </Popup>
</div>
`;
